librz/util/thread_pool.c

```diff
@@ -0,0 +1,201 @@
+ *
+ * \param  max_threads  The maximum number of threads needed in the pool
+ * \return RzThreadPool The RzThreadPool structure
+ */
+RZ_API RZ_OWN RzThreadPool *rz_th_pool_new(size_t max_threads) {
+	RzThreadPool *pool = RZ_NEW0(RzThreadPool);
+	if (!pool) {
+		return NULL;
+	}
+
+	size_t cores = rz_th_physical_core_number();
+	if (max_threads) {
+		cores = RZ_MIN(cores, max_threads);
+	}
+
+	pool->size = cores;
+	pool->threads = RZ_NEWS0(RzThread *, cores);
+	if (!pool->threads) {
+		free(pool);
+		return NULL;
+	}
+
+	return pool;
+}
+
+/**
+ * \brief Kills (and frees) the threads and frees the RzThreadPool struct
+ *
+ * \param RzThreadPool *The thread pool to free
+ */
+RZ_API void rz_th_pool_free(RZ_NULLABLE RzThreadPool *pool) {
+	if (!pool) {
+		return;
+	}
+	for (ut32 i = 0; i < pool->size; ++i) {
+		if (pool->threads[i]) {
+			rz_th_free(pool->threads[i]);
+			pool->threads[i] = NULL;
+		}
+	}
+	free(pool->threads);
+	free(pool);
+}
+
+/**
+ * \brief Adds a thread to the thread pool
+ *
+ * \param  RzThreadPool  The thread pool where to add the thread
+ * \param  RzThread      The thread to add to the pool
+ * \return true if a slot is found, false otherwise
+ */
+RZ_API bool rz_th_pool_add_thread(RZ_NONNULL RzThreadPool *pool, RZ_NONNULL RzThread *thread) {
+	rz_return_val_if_fail(pool && thread, false);
+	for (ut32 i = 0; i < pool->size; ++i) {
+		if (!pool->threads[i]) {
+			RZ_LOG_DEBUG("thread: thread %u added\n", i);
+			pool->threads[i] = thread;
+			return true;
+		}
+	}
+	return false;
+}
+
+/**
+ * \brief  Returns the n-th thread in the thread pool.
+ *
+ * \param  pool   The thread pool to use
+ * \param  index  The index of the thread to get
+ *
+ * \return Returns the pointer of the n-th thread in the thread pool.
+ */
+RZ_API RZ_BORROW RzThread *rz_th_pool_get_thread(RZ_NONNULL RzThreadPool *pool, size_t index) {
+	rz_return_val_if_fail(pool && index < pool->size, NULL);
+	return pool->threads[index];
+}
+
+/**
+ * \brief Waits the end of all the threads in the thread pool
+ *
+ * \param  RzThreadPool The thread pool to wait for
+ *
+ * \return true if managed to wait all threads, otherwise false
+ */
+RZ_API bool rz_th_pool_wait(RZ_NONNULL RzThreadPool *pool) {
+	rz_return_val_if_fail(pool, false);
+	bool has_exited = true;
+	for (ut32 i = 0; i < pool->size; ++i) {
+		if (pool->threads[i]) {
+			RZ_LOG_DEBUG("thread: waiting for thread %u\n", i);
+			has_exited = has_exited && rz_th_wait(pool->threads[i]);
+		}
+	}
+	return has_exited;
+}
+
+/**
+ * \brief Force-stops all threads in the thread pool
+ *
+ * \param  pool  The thread pool to kill
+ *
+ * \return true if managed to kill all threads, otherwise false
+ */
+RZ_API bool rz_th_pool_kill(RZ_NONNULL RzThreadPool *pool) {
+	rz_return_val_if_fail(pool, false);
+	bool has_exited = false;
+	for (ut32 i = 0; i < pool->size; ++i) {
+		if (pool->threads[i]) {
+			RZ_LOG_DEBUG("thread: killing thread %u\n", i);
+			rz_th_kill(pool->threads[i]);
```

Comment:
Forcefully killing threads is not a good idea and makes the entire process enter an undefined state. It would be best to have a way to signal the thread pool thread to exit normally.

Reply:
that is up to the person using the code. ofc, you should not normally do this.